# Mini Project: Transfer Learning with Keras

Transfer learning is a machine learning technique where a model trained on one task is used as a starting point to solve a different but related task. Instead of training a model from scratch, transfer learning leverages the knowledge learned from the source task and applies it to the target task. This approach is especially useful when the target task has limited data or computational resources.

In transfer learning, the pre-trained model, also known as the "base model" or "source model," is typically trained on a large dataset and a more general problem (e.g., image classification on ImageNet, a vast dataset with millions of labeled images). The knowledge learned by the base model in the form of feature representations and weights captures common patterns and features in the data.

To perform transfer learning, the following steps are commonly followed:

1. Pre-training: The base model is trained on a source task using a large dataset, which can take a considerable amount of time and computational resources.

2. Feature Extraction: After pre-training, the base model is used as a feature extractor. The last few layers (classifier layers) of the model are discarded, and the remaining layers (feature extraction layers) are retained. These layers serve as feature extractors, producing meaningful representations of the data.

3. Fine-tuning: The feature extraction layers and sometimes some of the earlier layers are connected to a new set of layers, often called the "classifier layers" or "task-specific layers." These layers are randomly initialized, and the model is trained on the target task with a smaller dataset. The weights of the base model can be frozen during fine-tuning, or they can be allowed to be updated with a lower learning rate to fine-tune the model for the target task.

Transfer learning has several benefits:

1. Reduced training time and resource requirements: Since the base model has already learned generic features, transfer learning can save time and resources compared to training a model from scratch.

2. Improved generalization: Transfer learning helps the model generalize better to the target task, especially when the target dataset is small and dissimilar from the source dataset.

3. Better performance: By starting from a model that is already trained on a large dataset, transfer learning can lead to better performance on the target task, especially in scenarios with limited data.

4. Effective feature extraction: The feature extraction layers of the pre-trained model can serve as powerful feature extractors for different tasks, even when the task domains differ.

Transfer learning is commonly used in various domains, including computer vision, natural language processing (NLP), and speech recognition, where pre-trained models are fine-tuned for specific applications like object detection, sentiment analysis, or speech-to-text.

In this mini-project you will perform fine-tuning using Keras with a pre-trained VGG16 model on the CIFAR-10 dataset.

First, import all the libraries you'll need.

In [2]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, GlobalAveragePooling2D
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split

The CIFAR-10 dataset is a widely used benchmark dataset in the field of computer vision and machine learning. It stands for the "Canadian Institute for Advanced Research 10" dataset. CIFAR-10 was created by researchers at the CIFAR institute and was originally introduced as part of the Neural Information Processing Systems (NIPS) 2009 competition.

The dataset consists of 60,000 color images, each of size 32x32 pixels, belonging to ten different classes. Each class contains 6,000 images. The ten classes in CIFAR-10 are:

1. Airplane
2. Automobile
3. Bird
4. Cat
5. Deer
6. Dog
7. Frog
8. Horse
9. Ship
10. Truck

The images are evenly distributed across the classes, making CIFAR-10 a balanced dataset. The dataset is divided into two sets: a training set and a test set. The training set contains 50,000 images, while the test set contains the remaining 10,000 images.

CIFAR-10 is often used for tasks such as image classification, object recognition, and transfer learning experiments. The relatively small size of the images and the variety of classes make it a challenging dataset for training machine learning models, especially deep neural networks. It also serves as a good dataset for teaching and learning purposes due to its manageable size and straightforward class labels.

Here are your tasks:

1. Load the CIFAR-10 dataset after referencing the documentation [here](https://keras.io/api/datasets/cifar10/).
2. Normalize the pixel values so they're all in the range [0, 1].
3. Apply One Hot Encoding to the train and test labels using the [to_categorical](https://www.tensorflow.org/api_docs/python/tf/keras/utils/to_categorical) function.
4. Further split the the training data into training and validation sets using [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html). Use only 10% of the data for validation.  

In [3]:
# Load the CIFAR-10 dataset
import keras
(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
assert x_train.shape == (50000, 32, 32, 3)
assert x_test.shape == (10000, 32, 32, 3)
assert y_train.shape == (50000, 1)
assert y_test.shape == (10000, 1)

170498071/170498071 [==============================] - 4s 0us/step


In [4]:
# Normalize the pixel values to [0, 1]
X_train, X_test = x_train / 255.0, x_test / 255.0

In [5]:
# One-hot encode the labels
train_labels = to_categorical(y_train, num_classes=10)
test_labels = to_categorical(y_test, num_classes=10)

In [6]:
# Split the data into training and validation sets (further from the training set)
train_images, val_images, train_labels, val_labels = train_test_split(
    X_train, train_labels, test_size=0.1, random_state=42
)

VGG16 (Visual Geometry Group 16) is a deep convolutional neural network architecture that was developed by the Visual Geometry Group at the University of Oxford. It was proposed by researchers Karen Simonyan and Andrew Zisserman in their paper titled "Very Deep Convolutional Networks for Large-Scale Image Recognition," which was presented at the International Conference on Learning Representations (ICLR) in 2015.

The VGG16 architecture gained significant popularity for its simplicity and effectiveness in image classification tasks. It was one of the pioneering models that demonstrated the power of deeper neural networks for visual recognition tasks.

Key characteristics of the VGG16 architecture:

1. Architecture: VGG16 consists of a total of 16 layers, hence the name "16." These layers are stacked one after another, forming a deep neural network.

2. Convolutional Layers: The main building blocks of VGG16 are the convolutional layers. It primarily uses 3x3 convolutional filters throughout the network, which allows it to capture local features effectively.

3. Max Pooling: After each set of convolutional layers, VGG16 applies max-pooling layers with 2x2 filters and stride 2, which halves the spatial dimensions (width and height) of the feature maps and reduces the number of parameters.

4. Fully Connected Layers: Towards the end of the network, VGG16 has fully connected layers that act as a classifier to make predictions based on the learned features.

5. Activation Function: The network uses the Rectified Linear Unit (ReLU) activation function for all hidden layers, which helps with faster convergence during training.

6. Number of Filters: The number of filters in each convolutional layer is relatively small compared to more recent architectures like ResNet or InceptionNet. However, stacking multiple layers allows VGG16 to learn complex hierarchical features.

7. Output Layer: The output layer consists of 1000 units, corresponding to 1000 ImageNet classes. VGG16 was originally trained on the large-scale ImageNet dataset, which contains millions of images from 1000 different classes.

VGG16 was instrumental in showing that increasing the depth of a neural network can significantly improve its performance on image recognition tasks. However, the main drawback of VGG16 is its high number of parameters, making it computationally expensive and memory-intensive to train. Despite this limitation, VGG16 remains an essential benchmark architecture and has paved the way for even deeper and more efficient models in the field of computer vision, such as ResNet, DenseNet, and EfficientNet.

Here are your tasks:

1. Load [VGG16](https://keras.io/api/applications/vgg/#vgg16-function) as a base model. Make sure to exclude the top layer.
2. Freeze all the layers in the base model. We'll be using these weights as a feature extraction layer to forward to layers that are trainable.

In [7]:
# Load the pre-trained VGG16 model (excluding the top classifier)
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))

# Display a summary of the base model
base_model.summary()

58889256/58889256 [==============================] - 0s 0us/step
Model: "vgg16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 32, 32, 3)]       0         
                                                                 
 block1_conv1 (Conv2D)       (None, 32, 32, 64)        1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 32, 32, 64)        36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 16, 16, 64)        0         
                                                                 
 block2_conv1 (Conv2D)       (None, 16, 16, 128)       73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 16, 16, 128)       147584    
                                                              

In [8]:
# Freeze the layers in the base model
for layer in base_model.layers:
    layer.trainable = False

# Display a summary of the base model
base_model.summary()

Model: "vgg16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 32, 32, 3)]       0         
                                                                 
 block1_conv1 (Conv2D)       (None, 32, 32, 64)        1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 32, 32, 64)        36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 16, 16, 64)        0         
                                                                 
 block2_conv1 (Conv2D)       (None, 16, 16, 128)       73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 16, 16, 128)       147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 8, 8, 128)         0     

Now, we'll add some trainable layers to the base model.

1. Using the base model, add a [GlobalAveragePooling2D](https://keras.io/api/layers/pooling_layers/global_average_pooling2d/) layer, followed by a [Dense](https://keras.io/api/layers/core_layers/dense/) layer of length 256 with ReLU activation. Finally, add a classification layer with 10 units, corresponding to the 10 CIFAR-10 classes, with softmax activation.
2. Create a Keras [Model](https://keras.io/api/models/model/) that takes in approproate inputs and outputs.

In [9]:
# Add a global average pooling layer

from tensorflow.keras.models import Sequential

# Create a Sequential model
custom_model = Sequential()

# Add the VGG16 base model (frozen) to the custom model
custom_model.add(base_model)

# Now add GlobalAveragePooling2D layer to the Sequential model
custom_model.add(GlobalAveragePooling2D())

In [10]:
# Add a fully connected layer with 256 units and ReLU activation
custom_model.add(Dense(256, activation='relu'))

In [11]:
# Add the final classification layer with 10 units (for CIFAR-10 classes) and softmax activation
custom_model.add(Dense(10, activation='softmax'))

In [12]:
# Create the fine-tuned model

# Display a summary of the custom model before compiling
custom_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 vgg16 (Functional)          (None, 1, 1, 512)         14714688  
                                                                 
 global_average_pooling2d (  (None, 512)               0         
 GlobalAveragePooling2D)                                         
                                                                 
 dense (Dense)               (None, 256)               131328    
                                                                 
 dense_1 (Dense)             (None, 10)                2570      
                                                                 
Total params: 14848586 (56.64 MB)
Trainable params: 133898 (523.04 KB)
Non-trainable params: 14714688 (56.13 MB)
_________________________________________________________________


With your model complete it's time to train it and assess its performance.

1. Compile your model using an appropriate loss function. Feel free to play around with the optimizer, but a good starting optimizer might be Adam with a learning rate of 0.001.
2. Fit your model on the training data. Use the validation data to print the accuracy for each epoch. Try training for 10 epochs. Note, training can take a few hours so go ahead and grab a cup of coffee.

**Optional**: See if you can implement an [Early Stopping](https://keras.io/api/callbacks/early_stopping/) criteria as a callback function.

In [13]:
# Compile the model
custom_model.compile(optimizer=Adam(learning_rate=0.0001), loss='categorical_crossentropy', metrics=['accuracy'])

In [14]:
# Train the model

from tensorflow.keras.callbacks import EarlyStopping

# Define Early Stopping callback
early_stopping = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

# Fit the model with Early Stopping callback
history = custom_model.fit(
    train_images, train_labels,
    epochs=10,
    validation_data=(val_images, val_labels),
    callbacks=[early_stopping],
)

Epoch 1/10
1407/1407 [==============================] - 18s 10ms/step - loss: 1.6689 - accuracy: 0.4348 - val_loss: 1.4181 - val_accuracy: 0.5094
Epoch 2/10
1407/1407 [==============================] - 13s 10ms/step - loss: 1.3595 - accuracy: 0.5347 - val_loss: 1.3145 - val_accuracy: 0.5438
Epoch 3/10
1407/1407 [==============================] - 14s 10ms/step - loss: 1.2806 - accuracy: 0.5600 - val_loss: 1.2631 - val_accuracy: 0.5600
Epoch 4/10
1407/1407 [==============================] - 14s 10ms/step - loss: 1.2362 - accuracy: 0.5750 - val_loss: 1.2371 - val_accuracy: 0.5690
Epoch 5/10
1407/1407 [==============================] - 14s 10ms/step - loss: 1.2045 - accuracy: 0.5871 - val_loss: 1.2136 - val_accuracy: 0.5802
Epoch 6/10
1407/1407 [==============================] - 14s 10ms/step - loss: 1.1811 - accuracy: 0.5934 - val_loss: 1.2000 - val_accuracy: 0.5802
Epoch 7/10
1407/1407 [==============================] - 14s 10ms/step - loss: 1.1617 - accuracy: 0.5996 - val_loss: 1.1816 -

With your model trained, it's time to assess how well it performs on the test data.

1. Use your trained model to calculate the accuracy on the test set. Is the model performance better than random?
2. Experiment! See if you can tweak your model to improve performance.  

In [15]:
# Evaluate the model on the test set
test_loss, test_accuracy = custom_model.evaluate(X_test, test_labels)
print(f'Test Accuracy: {test_accuracy}')

313/313 [==============================] - 3s 10ms/step - loss: 1.1803 - accuracy: 0.5882
Test Accuracy: 0.5881999731063843


We see that the trained model does reasonably well with an accuracy of about .58; a random guess would yield an accuracy of .1 since there are 10 classes to choose from.


How can we get better results from our model? The easiest thing to try would be to just train the model for more epochs since we saw val loss continue to go down and val accuracy continue to go up over the 10 epochs fitting to the training data.


In [16]:
# Fit the model for another 30 epochs with Early Stopping callback
history = custom_model.fit(
    train_images, train_labels,
    epochs=30,
    validation_data=(val_images, val_labels),
    callbacks=[early_stopping],
)

Epoch 1/30
1407/1407 [==============================] - 14s 10ms/step - loss: 1.1093 - accuracy: 0.6174 - val_loss: 1.1481 - val_accuracy: 0.6008
Epoch 2/30
1407/1407 [==============================] - 14s 10ms/step - loss: 1.0984 - accuracy: 0.6207 - val_loss: 1.1430 - val_accuracy: 0.6034
Epoch 3/30
1407/1407 [==============================] - 15s 11ms/step - loss: 1.0882 - accuracy: 0.6252 - val_loss: 1.1504 - val_accuracy: 0.6002
Epoch 4/30
1407/1407 [==============================] - 14s 10ms/step - loss: 1.0804 - accuracy: 0.6263 - val_loss: 1.1367 - val_accuracy: 0.6038
Epoch 5/30
1407/1407 [==============================] - 14s 10ms/step - loss: 1.0719 - accuracy: 0.6303 - val_loss: 1.1294 - val_accuracy: 0.6048
Epoch 6/30
1407/1407 [==============================] - 14s 10ms/step - loss: 1.0636 - accuracy: 0.6339 - val_loss: 1.1282 - val_accuracy: 0.6036
Epoch 7/30
1407/1407 [==============================] - 14s 10ms/step - loss: 1.0562 - accuracy: 0.6360 - val_loss: 1.1237 -

We see our early stopping function kicked in after 20 epochs, meaning that the val loss started tapering off once the val accuracy reached about .62.

In [17]:
# Evaluate the model on the test set
test_loss, test_accuracy = custom_model.evaluate(X_test, test_labels)
print(f'Test Accuracy: {test_accuracy}')

313/313 [==============================] - 3s 9ms/step - loss: 1.1087 - accuracy: 0.6169
Test Accuracy: 0.6169000267982483


We see a little improvement on the test set from training for more epochs. The accuracy went from about .58 to about .61. The output of the training process suggests that overfitting had already started to happen since the accuracy on the training set was still improving while the validation accuracy and loss began to taper off.

So what other things can we try to get better accuracy from our custom model? Two ideas are to use data augmentation to make the training dataset bigger and more diverse, and to unfreeze the earlier layers during training with a low learning rate.
What I did to try data augmentation didn't work and didn't get saved; it would either yield worse results or when I tried making the training dataset larger the runtime would crash because it ran out of RAM. I may try a different approach for data augmentation (maybe I did not do something right) but for now I want to try something else: unfreeze earlier layers and train model again with smaller learning rate.   

Thus I'll try fine tuning the already trained model by unfreezing the earlier layers.

In [18]:
# Unfreeze the layers in the base model
for layer in custom_model.layers:
    layer.trainable = True

# Recompile the model
custom_model.compile(optimizer=Adam(learning_rate=0.0001), loss='categorical_crossentropy', metrics=['accuracy'])

# Continue training the model with unfrozen layers
history = custom_model.fit(
    train_images, train_labels,
    epochs=10,
    validation_data=(val_images, val_labels),
    callbacks=[early_stopping],
)

Epoch 1/10
1407/1407 [==============================] - 51s 31ms/step - loss: 0.8458 - accuracy: 0.7090 - val_loss: 0.6513 - val_accuracy: 0.7774
Epoch 2/10
1407/1407 [==============================] - 44s 31ms/step - loss: 0.4986 - accuracy: 0.8285 - val_loss: 0.4850 - val_accuracy: 0.8358
Epoch 3/10
1407/1407 [==============================] - 43s 31ms/step - loss: 0.3586 - accuracy: 0.8780 - val_loss: 0.5367 - val_accuracy: 0.8214
Epoch 4/10
1407/1407 [==============================] - 44s 31ms/step - loss: 0.2598 - accuracy: 0.9115 - val_loss: 0.4786 - val_accuracy: 0.8470
Epoch 5/10
1407/1407 [==============================] - 44s 31ms/step - loss: 0.1891 - accuracy: 0.9353 - val_loss: 0.5214 - val_accuracy: 0.8404
Epoch 6/10
1407/1407 [==============================] - 44s 31ms/step - loss: 0.1385 - accuracy: 0.9534 - val_loss: 0.5164 - val_accuracy: 0.8550
Epoch 7/10
1407/1407 [==============================] - 44s 31ms/step - loss: 0.1053 - accuracy: 0.9648 - val_loss: 0.5965 -

Val accuracy is at 85% now, so we gained a lot from unfreezing all the layers of the model.

In [19]:
# Evaluate the model on the test set
test_loss, test_accuracy = custom_model.evaluate(X_test, test_labels)
print(f'Test Accuracy: {test_accuracy}')

313/313 [==============================] - 3s 8ms/step - loss: 0.4894 - accuracy: 0.8443
Test Accuracy: 0.8442999720573425


We see that the accuracy on the test set is now 84%, which is quite a great improvement from the 61% accuracy the model predicted with prior to training the earlier layers of the network.